### PR TITLE
chore: Attachment use compose()

### PR DESCRIPTION
**Pull request by layershifter**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ yarn change`

# BREAKING CHANGES

### `slotClassNames` removal

`slotClassNames` were removed from `Attachment`, please use classnames on matching components instead.

```tsx
// Before
Attachment.slotClassNames.action
// After
AttachmentAction.className
```

### `styles` were moved to components from slots

- styles for `action`, `description`, `icon` & `header` slots were moved to separate components
  - as these components are separate they have own props for styles
- styles for `content` slots were moved to separate `AttachmentBody`
- styles for `progress` slot are not separated anymore and should be used on `root` in `attachmentStyles`

#### Before

```ts
const attachmentStyles: ComponentSlotStylesPrepared<AttachmentStylesProps, AttachmentVariables> = {
  content: () => ({ display: 'flex' }),
  progress: () => ({ display: 'flex' }),

  action: () => ({ display: 'flex' }),
  description: () => ({ display: 'flex' }),
  icon: () => ({ display: 'flex' }),
  header: () => ({ display: 'flex' }),
}
```

#### After

```tsx
const attachmentStyles: ComponentSlotStylesPrepared<AttachmentStylesProps, AttachmentVariables> = {
  root: () => ({
    '& .ui-attachment__progress': {
      display: 'flex'
   },
};

const attachmentBodyStyles: ComponentSlotStylesPrepared<AttachmentBodyStylesProps, AttachmentVariables> = {
  root: ({ variables: v }): ICSSInJSStyle => ({
    display: 'flex'
  }),
};

const attachmentIconStyles: ComponentSlotStylesPrepared<AttachmentIconStylesProps, AttachmentVariables> = {
  root: ({ variables: v }): ICSSInJSStyle => ({
    display: 'flex'
  }),
};

const attachmentHeaderStyles: ComponentSlotStylesPrepared<AttachmentHeaderStylesProps, AttachmentVariables> = {
  root: ({ variables: v }): ICSSInJSStyle => ({
    display: 'flex'
  }),
};
```

#### Description of changes

This PR updates `Attachment` to use instead `Box`es composed components.
Adds `body` slot to `Attachment`.

#### Performance

Due enabled caching performance for rerenders was improved improved almost twice 🚀 

image

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/fluentui/pull/12674)

### CODE REVIEW

File: packages/fluentui/react-bindings/src/telemetry/useTelemetry.ts [hunk 1]
```diff
@@ -34,8 +34,9 @@ export const getTelemetry = (displayName: string, telemetry: Telemetry | undefin
 
 const useTelemetry = (displayName: string, telemetry: Telemetry | undefined): UseTelemetryResult => {
   const composeOptions = useComposeOptions();
+  const telemetryName = composeOptions?.displayNames[composeOptions?.displayNames.length - 1] || displayName;
```
Comment (layershifter) [inline packages/fluentui/react-bindings/src/telemetry/useTelemetry.ts:37]: Oops, the actual name is the last one

File: packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx [hunk 2]
```diff
@@ -122,29 +124,17 @@ const Attachment: React.FC<WithAsProp<AttachmentProps>> &
...
       {(header || description) && (
-        <div className={classes.content}>
-          {Text.create(header, {
-            defaultProps: () => ({ styles: resolvedStyles.header }),
-          })}
-
-          {Text.create(description, {
-            defaultProps: () => ({ styles: resolvedStyles.description }),
-          })}
+        <div className="ui-attachment__content">
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx:130]: Just curious, why we don't have `AttachmentBody` component that will have the header and description props and we would create those inside that component?

File: packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx [hunk 3]
```diff
@@ -168,13 +155,18 @@ Attachment.propTypes = {
   actionable: PropTypes.bool,
   description: customPropTypes.itemShorthand,
   header: customPropTypes.itemShorthand,
-  icon: customPropTypes.shorthandAllowingChildren,
+  icon: customPropTypes.itemShorthand,
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx:158]: Why was this changed? Does this means that we are changing the API here? I would expect that even if we have different component, to keep the same API..

File: packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx [hunk 4]
```diff
@@ -122,29 +124,17 @@ const Attachment: React.FC<WithAsProp<AttachmentProps>> &
...
-          iconOnly: true,
-          text: true,
-          styles: resolvedStyles.action,
-          className: Attachment.slotClassNames.action,
-        }),
-      })}
-      {!_.isNil(progress) && <div className={classes.progress} style={{ width: `${progress}%` }} />}
+
+      {AttachmentAction.create(action)}
+      {!_.isNil(progress) && <div className="ui-attachment__progress" style={{ width: `${progress}%` }} />}
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx:137]: Why we don't have `AttachmentProgress` component? I understand that you are just mapping shorthands to components, so I'm just interested whether there is a solid reason for it not being one

File: packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx [hunk 5]
```diff
@@ -122,29 +124,17 @@ const Attachment: React.FC<WithAsProp<AttachmentProps>> &
...
-        defaultProps: () => ({
-          iconOnly: true,
-          text: true,
-          styles: resolvedStyles.action,
-          className: Attachment.slotClassNames.action,
-        }),
-      })}
-      {!_.isNil(progress) && <div className={classes.progress} style={{ width: `${progress}%` }} />}
+
+      {AttachmentAction.create(action)}
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx:136]: It's interesting that none of the sub-components need any of the parent's props :)

File: packages/fluentui/react-northstar/test/specs/commonTests/isConformant.tsx [hunk 6]
```diff
@@ -65,6 +66,8 @@ export default function isConformant(
   const { throwError } = helpers('isConformant', Component);
 
   const componentType = typeof Component;
+  // composed components store `handledProps` under config
+  const handledProps = Component.handledProps || (Component as ComposedComponent).fluentComposeConfig?.handledProps;
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/test/specs/commonTests/isConformant.tsx:70]: Shuoldn't this be:
```
const handledProps = (Component as ComposedComponent).fluentComposeConfig?.handledProps || Component.handledProps;
```

File: packages/fluentui/react-northstar/src/themes/teams/components/Attachment/attachmentStyles.ts [hunk 7]
```diff
@@ -33,6 +33,21 @@ const attachmentStyles: ComponentSlotStylesPrepared<AttachmentStylesProps, Attac
 
       ...borderFocusStyles,
 
+      '& .ui-attachment__content': {
+        flex: 1,
+      },
+
+      '& .ui-attachment__progress': {
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/themes/teams/components/Attachment/attachmentStyles.ts:40]: It's not obvious why this is implemented with selector. I would vote for a separate component, to keep it consistent with other slots..

File: packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx (hunk 3 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx:158]: Merge issue, should be reverted

File: packages/fluentui/react-northstar/test/specs/commonTests/isConformant.tsx (hunk 6 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/test/specs/commonTests/isConformant.tsx:70]: Hm... Good point. Actually these is no difference as there is no `handledProps` in static props
